### PR TITLE
feat: rollback vm state on msg failure

### PR DESCRIPTION
**Pull request by icorderi**
### Motivation

Changes on the state tree during the execution of a msg need to be rolled back if there is a runtime abort raised during the execution. 

### Proposed changes

- cleanup the different places we were handling errors to adjust to the latest send protection.
- impl rollback on state 
- add rollback calls if an abort is raised 

Closes issue #3718

### CODE REVIEW

File: internal/pkg/vm/internal/vmcontext/invocation_context.go [hunk 1]
```diff
@@ -218,7 +260,11 @@ func (ctx *invocationContext) resolveTarget(target address.Address) (*actor.Acto
 		}
 
 		newCtx := newInvocationContext(ctx.rt, newMsg, nil, ctx.gasTank, ctx.randSource)
-		newCtx.invoke()
+		_, code := newCtx.invoke()
+		if code != exitcode.Ok {
```
Comment (frrist) [inline internal/pkg/vm/internal/vmcontext/invocation_context.go]: Replace with https://github.com/filecoin-project/specs-actors/blob/master/actors/runtime/exitcode/reserved.go#L11

File: internal/pkg/vm/internal/vmcontext/vmcontext.go [hunk 2]
```diff
@@ -330,8 +327,13 @@ func (vm *VM) applyImplicitMessage(imsg internalMessage, rnd crypto.RandomnessSo
 
 	// 3. build context
 	ctx := newInvocationContext(vm, imsg, fromActor, &gasTank, rnd)
+
 	// 4. invoke message
-	return ctx.invoke(), nil
+	ret, code := ctx.invoke()
+	if code != exitcode.Ok {
```
Comment (frrist) [inline internal/pkg/vm/internal/vmcontext/vmcontext.go]: https://github.com/filecoin-project/specs-actors/blob/master/actors/runtime/exitcode/reserved.go#L11

File: internal/pkg/vm/state/state.go [hunk 3]
```diff
@@ -138,6 +139,21 @@ func (st *State) Commit(ctx context.Context) (Root, error) {
 	return st.root, nil
 }
 
+// Rollback discards all uncommitted changes.
+func (st *State) Rollback(ctx context.Context) error {
+	// load the original root node again
+	rootNode, err := hamt.LoadNode(ctx, st.store, st.root, hamt.UseTreeBitWidth(TreeBitWidth))
+
```
Comment (frrist) [inline internal/pkg/vm/state/state.go]: remove new line

File: internal/pkg/vm/internal/vmcontext/invocation_context.go [hunk 4]
```diff
@@ -279,6 +325,18 @@ type returnWrapper struct {
 	inner specsruntime.CBORMarshaler
 }
 
+func (r returnWrapper) ToCbor() ([]byte, error) {
+	if r.inner == nil {
+		return []byte{}, nil
+	}
+	b := bytes.Buffer{}
+	err := r.inner.MarshalCBOR(&b)
+	if err != nil {
```
Comment (frrist) [inline internal/pkg/vm/internal/vmcontext/invocation_context.go]: nit: single line

**Comment by anorth** [comment]
This came out nicely.

**Pull request merged**